### PR TITLE
Fix/#692 transitive selfrefs

### PR DESCRIPTION
**Pull request by odersky**
This was a longer thread than I expected. Review by @DarkDimius

### CODE REVIEW

File: src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -3181,14 +3182,23 @@ object Types {
...
+    extends TypeError(
+      s"malformed type: $pre is not a legal prefix for $denot because it contains abstract type member${if (absMembers.size == 1) "" else "s"} ${absMembers.mkString(", ")}")
+
+  class MissingType(pre: Type, name: Name)(implicit ctx: Context) extends TypeError(
+    i"""cannot resolve reference to type $pre.$name
+       |the classfile defining the type might be missing from the classpath${otherReason(pre)}""".stripMargin)
+
+  private def otherReason(pre: Type)(implicit ctx: Context): String = pre match {
+    case pre: ThisType if pre.givenSelfType.exists =>
+      i"\nor the self type of $pre might not contain all transitive dependencies"
```
Comment (smarter) [inline src/dotty/tools/dotc/core/Types.scala]: Typo: "nor" -> "or"

Comment (DarkDimius) [inline src/dotty/tools/dotc/core/Types.scala]: isn't `\n` here an escaped "end of line" character?

Comment (smarter) [inline src/dotty/tools/dotc/core/Types.scala]: Ah, you're right, I misread that.

**Comment by DarkDimius** [comment]
LGTM, needs rebasing.

**Comment by odersky** [comment]
Rebased to master

**Comment by odersky** [comment]
Closes #690

**Pull request merged**